packages/react-jss/src/create-use-styles.js

```diff
@@ -0,0 +1,101 @@
+// @flow
+
+import React from 'react'
+import {ThemeContext as DefaultThemeContext} from 'theming'
+
+import {JssContext} from './JssContext'
+import {
+  createStaticSheet,
+  addDynamicRules,
+  removeDynamicRules,
+  updateDynamicRules
+} from './utils/sheets'
+import {getIndex} from './utils/index-counter'
+import type {HookOptions, Styles} from './types'
+import {unmanageSheet, manageSheet} from './utils/managers'
+import {getSheetClasses} from './utils/get-sheet-classes'
+
+const noTheme = {}
+
+function createUseStyles<Theme: {}>(styles: Styles<Theme>, options?: HookOptions<Theme> = {}) {
+  const {index = getIndex(), theming, name = 'Hook', ...sheetOptions} = options
+  const ThemeContext = (theming && theming.context) || DefaultThemeContext
+
+  return (data: any) => {
+    const jssContext = React.useContext(JssContext)
+    // $FlowFixMe
+    const theme: Theme = typeof styles === 'function' ? React.useContext(ThemeContext) : noTheme
+
+    // When the theme or the context changes we create a new sheet
+    const sheet = React.useMemo(
```

Comment:
React doesn't provide any guarantee about the underlying caching of useMemo. The cache might be clear at any time. You probably don't want to use it for anything involving `manager.add(options.theme, sheet)`. It "might" leak.

Reply:
Also, the sheet attach should be performed in an effect. React can trigger a render to later ditch the work started. If the render has inserted a sheet into the page, no teardown will be triggered, the sheet will never be removed. It leaks. That's 💯 guarantee, we have this problem on MUI side. We didn't refactor the code yet.

Reply:
So we should use a combination of `useState` with an initial state function and `useLayoutEffect` to flush changes to the DOM before an update happens.

Actually, on SSR we only need to add it to the registry and not manage the sheet, so we could move the managing of the sheet into a `useLayoutEffect` where we could also clean it up. We need to make sure that there are no flashes of unstyled content then.

Reply:
useLayoutEffect will not cause a FOUC